### ISSUE TITLE
Question about reproduction performance.

### ISSUE THREAD

**Issue by mongeoroo** [yuhui-zh15/C3#1]
Hello, 

I appreciate your well-structured code base and nice work.

I have tried reproduction for C3 performance reported in Table 2.

However, I got results slightly lower than table 2.

| **Test Metric** | **DataLoader 0** |
|-----------------|------------------|
| test/Bleu_1     | 0.6807827949523926 |
| test/Bleu_2     | 0.5076174736022949 |
| test/Bleu_3     | 0.3580711781978607 |
| test/Bleu_4     | 0.25035572052001953 |
| test/CIDEr      | 0.8898507356643677 |
| test/METEOR     | 0.2402649223804474 |
| test/ROUGE_L    | 0.5052425265312195 |
| test/SPICE      | 0.17114298045635223 |

I assigned the noise level as sqrt(0.016).  And I test the weight of 8 epoch trained model because it showed the highest METEOR score (epoch=8-step=15939.ckpt).

Can you share the configuration or some tips about this gap?

Below is my configuration file.

Thank you for your great job.

### Configuration File

```yaml
experiment_name: coco_stage1_unimodal_text_reconstruction_normed_mlp_remove_mean_add_gaussian_noise_level_0.12649_AdamW_seed_1234
lightning:
  trainer:
    devices: '8'
    max_epochs: 10
    lr: 2.0e-05
    precision: 16
  checkpoint_callback:
    monitor: val/meteor
    dirpath: ./data/c3/coco/ckpt/coco_stage1_unimodal_text_reconstruction_normed_mlp_remove_mean_add_gaussian_noise_level_0.12649_AdamW_seed_1234/2024_10_14_15_03_24
    save_last: true
    mode: max
    save_top_k: 1
  logger:
    save_dir: ./data/logger/
    name: coco_stage1_unimodal_text_reconstruction_normed_mlp_remove_mean_add_gaussian_noise_level_0.12649_AdamW_seed_1234/2024_10_14_15_03_24
    project: c3
    logger_type: WandbLogger
encoder:
  modality: vision
  embed_dim: 512
  clip_model_type: ViT-B/32
decoder:
  modality: language
  model: gpt2
model:
  mapping_type: mlp
  prefix_length: 10
  clip_size: 10
  is_rn: false
  normalize_prefix: true
  re_normalize_prefix: true
  subsample_val_test: false
data:
  dataset: coco
  seed: 1234
  train_split: train
  train_data_path: ./data/coco/oscar_split_ViT-B_32_train.pkl
  val_data_path: ./data/coco/oscar_split_ViT-B_32_val.pkl
  test_data_path: ./data/coco/oscar_split_ViT-B_32_test.pkl
  text_embed_mean_path: ./data/coco/normalized_text_embed_mean.pkl
  image_embed_mean_path: ./data/coco/normalized_image_embed_mean.pkl
  sample_frac: null
  remove_mean: true
  diffusion: null
  add_gaussian_noise: true
train:
  batch_size: 40
  num_workers: 8
  loss_fn:
    name: ce_caption_loss
  optimizer:
    name: AdamW
  scheduler:
    name: linear_schedule_with_warmup
    warmup_steps: 5000
hyperparameters: {}
val_eval: true
cross_modal_val: true
checkpoint: null
test_split: test
noise_level: 0.12649110640673517
extension: '2024_10_14_15_03_24'
seed: 1234
output_dir: ./data/output/coco_stage1_unimodal_text_reconstruction_normed_mlp_remove_mean_add_gaussian_noise_level_0.12649_AdamW_seed_1234/2024_10_14_15_03_24

**Comment by elaine-sui** [comment]
Thanks for your question! It looks like I uploaded an older config version. It's updated now with minimizing 'val_loss_epoch' as the objective. For reference, my best epoch for seed=1234 is epoch=9-step=141680.ckpt. 
Also, looking at the number of steps in your final epoch, it seems that you may have not used the entire training set. We utilize the combination of the "train" and "restval" splits for our training set.

**Comment by mongeoroo** [comment]
Thank you for your quick and kind reply!

and Thanks for your point about the number of steps. 

Actually, I trained my model with COCO train + restval set. The code said it's 

Number of images is 113287
Number of captions is 566747

I made the pickle file using "parse_coco.py"

In my opinion, the small iteration comes from 8 GPUs.

For me, 1 epoch is 1771 iteration. 1771 * 8 * 40 (batch) = 566,720

In the parse_coco.py it saves restval as train split too.

Also, I tested all of the epochs of the weight file. However I can't obtain the result reported in the paper.

The gap may come from the multi-GPU processing. I guess...

How many GPUs did you use?

**Comment by elaine-sui** [comment]
Oh I see. We only used one GPU

**Comment by mongeoroo** [comment]
Thank you.

After finishing single GPU training, I will tell you the results.

**Comment by mongeoroo** [comment]
I got the results for the single GPU training. 

But the performance is lower than the reported table. (I also used "epoch=9-step=141680.ckpt")

Can you give me any suspicious point that I have to check?

*** In my opinion, the code that you used for the paper is some what different from the github code because the val_loss doesn't calculate the loss so I can't use val_loss as weight saving standard. So, Could you update your code entirely?

or which gpt-2 model did you used? In your code, gpt-2 small model was utilized. 

| **Test Metric** | **DataLoader 0** |
|-----------------|------------------|
| test/Bleu_1     | 0.7020882452002859|
| test/Bleu_2     | 0.5106101994588795 |
| test/Bleu_3     | 0.35596235728968506 |
| test/Bleu_4     | 0.24375196756986564 |
| test/CIDEr      | 0.9127197201491876 |
| test/METEOR     | 0.2357643299408623 |
| test/ROUGE_L    | 0.4992969525145615 |
| test/SPICE      | 0.17161963967581528 |

**Comment by elaine-sui** [comment]
I looked through the code again and it's the exact same. And yes, we use gpt-2 small. When I ran my experiments, I tried checkpointing with max val/meteor and min val_epoch_loss (need to set val_eval = false for this), yielding the same checkpoint for seed 1234 for both metrics. I've uploaded my model checkpoint for the seed [here](https://drive.google.com/file/d/1e4Jk9d-ktQgxzi--P8l-XUhiCN1S6YI5/view?usp=sharing).

These are the exact results I get with this checkpoint:
![Screenshot 2024-10-15 at 5 22 59 PM](https://github.com/user-attachments/assets/1706bce3-c77e-46ee-8f26-0b855a165390)

Let me know if you can reproduce these results when running evaluation on this checkpoint.

You can also try re-training with the updated config (monitor min val_loss_epoch) and scripts (setting val_eval = False, cross_modal_val = False).

**Comment by mongeoroo** [comment]
I really appreciate your support.

I tried to load the weight you shared but I got "unexpected keys" error. like below.

I loaded the 'gpt2' model using the github code /image_captioning/src/models/builder.py.

In my opinion, it may occur because of the version of transformers library.

Can you check the transformer version? Then I'll change my transformer library version and try to implement the results.



Unexpected key(s) in state_dict: "model.model.transformer.h.0.attn.bias", "model.model.transformer.h.0.attn.masked_bias", "model.model.transformer.h.1.attn.bias", "model.model.transformer.h.1.attn.masked_bias", "model.model.transformer.h.2.attn.bias", "model.model.transformer.h.2.attn.masked_bias", "model.model.transformer.h.3.attn.bias", "model.model.transformer.h.3.attn.masked_bias", "model.model.transformer.h.4.attn.bias", "model.model.transformer.h.4.attn.masked_bias", "model.model.transformer.h.5.attn.bias", "model.model.transformer.h.5.attn.masked_bias", "model.model.transformer.h.6.attn.bias", "model.model.transformer.h.6.attn.masked_bias", "model.model.transformer.h.7.attn.bias", "model.model.transformer.h.7.attn.masked_bias", "model.model.transformer.h.8.attn.bias", "model.model.transformer.h.8.attn.masked_bias", "model.model.transformer.h.9.attn.bias", "model.model.transformer.h.9.attn.masked_bias", "model.model.transformer.h.10.attn.bias", "model.model.transformer.h.10.attn.masked_bias", "model.model.transformer.h.11.attn.bias", "model.model.transformer.h.11.attn.masked_bias".

**Comment by elaine-sui** [comment]
It looks like it might be due to the tokenizer. My original code used GPT2Tokenizer and not AutoTokenizer, which I forgot to change back. The repo is edited for this now.

**Comment by mongeoroo** [comment]
I really appreciate your support.

Now I'm training the model with single GPU, Transformer version 4.28.1, and GPT2 tokenizer.

After finishing the training, I will show you the results.

**Comment by mongeoroo** [comment]
Thank you once again for your support.

I completed two tasks:

First, I tested the weights you shared and obtained results almost identical to those in Table 2:

```plaintext
"Bleu_1": 0.7092585089140861,
"Bleu_2": 0.5328332352624517,
"Bleu_3": 0.38609306041245295,
"Bleu_4": 0.27577497418256336,
"METEOR": 0.2499396847869458,
"ROUGE_L": 0.5196546083312271,
"CIDEr": 0.9317512777356817,
"SPICE": 0.1828019917910563
```

This confirms the reliability of my testing code. However, there is a slight discrepancy between these results and the figures you provided, suggesting potential issues with the generation code of the decoder. Given that the weights are strictly loaded on my decoder model and the architecture is presumed identical, I suspect the difference may lie in the sentence generation component.

Secondly, I trained the model with the specified settings (single GPU, Transformer version 4.28.1, and GPT2 tokenizer). Unfortunately, I observed no performance improvement over my initial implementation, even after experimenting with different seed values for training.

```plaintext
"Bleu_1": 0.6904916700528101,
"Bleu_2": 0.5124412888517843,
"Bleu_3": 0.3658407437141938,
"Bleu_4": 0.25807116099065225,
"METEOR": 0.2390900832270108,
"ROUGE_L": 0.5052022753719236,
"CIDEr": 0.8647266398700049,
"SPICE": 0.17097498640256317
```


This leads me to believe that the training process might differ from that described in your paper. One particular concern is the noise standard deviation. I noticed that the standard deviation used in your code is sqrt(0.016), identical to the one used in the 'capdec'. 

Could you clarify if this is the same value used in your training settings?

**Comment by elaine-sui** [comment]
Yes, the noise level is correct.

**Comment by mongeoroo** [comment]
Thank you for your support.

I finally get the results from the weight that I trained.

I don't know the reason. But for me, I can get the good result by setting the warmup step as 10,000.

So, single GPU, GPT2 tokenizer, and 2 times of warmup step improve the performance.

Thank you again.

```plaintext
Test metric             DataLoader 0
test/Bleu_1          0.7101796269416809
test/Bleu_2          0.5324719548225403
test/Bleu_3          0.3849830627441406
test/Bleu_4          0.27424734830856323
test/CIDEr           0.9282495164394607
test/METEOR          0.24859251081943512
test/ROUGE_L          0.5188136334631238
test/SPICE           0.1831031564703945
